dot/state/grandpa.go

```diff
@@ -0,0 +1,217 @@
+}
+
+// GetAuthorities returns the authorities for the given setID
+func (s *GrandpaState) GetAuthorities(setID uint64) ([]*types.GrandpaVoter, error) {
+	enc, err := s.db.Get(authoritiesKey(setID))
+	if err != nil {
+		return nil, err
+	}
+
+	r := &bytes.Buffer{}
+	_, _ = r.Write(enc)
+	v, err := types.DecodeGrandpaVoters(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+// setCurrentSetID sets the current set ID
+func (s *GrandpaState) setCurrentSetID(setID uint64) error {
+	buf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(buf, setID)
+	return s.db.Put(currentSetIDKey, buf[:])
+}
+
+// GetCurrentSetID retrieves the current set ID
+func (s *GrandpaState) GetCurrentSetID() (uint64, error) {
+	id, err := s.db.Get(currentSetIDKey)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(id) < 8 {
+		return 0, errors.New("invalid setID")
+	}
+
+	return binary.LittleEndian.Uint64(id), nil
+}
+
+// SetNextChange sets the next authority change
+func (s *GrandpaState) SetNextChange(authorities []*types.GrandpaVoter, number *big.Int) error {
+	currSetID, err := s.GetCurrentSetID()
+	if err != nil {
+		return err
+	}
+
+	nextSetID := currSetID + 1
+	err = s.setAuthorities(nextSetID, authorities)
+	if err != nil {
+		return err
+	}
+
+	err = s.setSetIDChangeAtBlock(nextSetID, number)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// IncrementSetID increments the set ID
+func (s *GrandpaState) IncrementSetID() error {
+	currSetID, err := s.GetCurrentSetID()
+	if err != nil {
+		return err
+	}
+
+	nextSetID := currSetID + 1
+	return s.setCurrentSetID(nextSetID)
+}
+
+// setSetIDChangeAtBlock sets a set ID change at a certain block
+func (s *GrandpaState) setSetIDChangeAtBlock(setID uint64, number *big.Int) error {
+	return s.db.Put(setIDChangeKey(setID), number.Bytes())
+}
+
+// GetSetIDChange returs the block number where the set ID was updated
+func (s *GrandpaState) GetSetIDChange(setID uint64) (*big.Int, error) {
+	num, err := s.db.Get(setIDChangeKey(setID))
+	if err != nil {
+		return nil, err
+	}
+
+	return big.NewInt(0).SetBytes(num), nil
+}
+
+// SetNextPause sets the next grandpa pause at the given block number
+func (s *GrandpaState) SetNextPause(number *big.Int) error {
+	return s.db.Put(pauseKey, number.Bytes())
+}
+
+// GetNextPause returns the block number of the next grandpa pause, nil if there is no upcoming pause
+func (s *GrandpaState) GetNextPause() (*big.Int, error) {
+	num, err := s.db.Get(pauseKey)
+	if err == chaindb.ErrKeyNotFound {
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return big.NewInt(0).SetBytes(num), nil
+}
+
+// SetNextResume sets the next grandpa resume at the given block number
+func (s *GrandpaState) SetNextResume(number *big.Int) error {
+	return s.db.Put(resumeKey, number.Bytes())
+}
+
+// GetNextResume returns the block number of the next grandpa resume, nil if there is no upcoming resume
+func (s *GrandpaState) GetNextResume() (*big.Int, error) {
```

Comment:
looks like this and `GetNextPause` can use the same codepath?

Reply:
one gets the `pauseKey` and one gets the `resumeKey`